stan/math/rev/mat/functor/cvodes_integrator.hpp

```diff
@@ -0,0 +1,376 @@
+        check_flag(CVodeSetMaxConvFails(cvode_mem_, max_conv_fails),
+                   "CVodeSetMaxConvFails");
+      }
+
+    public:
+      /**
+       * Construct CVODES integrator for an ODE model with initial
+       * state, initial time, flag for sensitivity calculation of the
+       * initial values and for the parameters, integrator options
+       * such as rel+abs tolerance and maximum number of steps. The
+       * integrator supports non-stiff (Adams-Moulton) and stiff (BDF)
+       * integration with optional stability detection.
+       *
+       * The integrator creates as output a vector of vector
+       * format. The outer vector is over the time-points where the
+       * solution is requested and the inner the states. The order of
+       * the states with is always
+       *
+       * \f[
+       * (y, \frac{\partial y}{\partial y_0}, \frac{\partial y}{\partial \theta}).
+       * \f]
+       *
+       * While the first N states correspond to y, all the remaining
+       * are optional and depend on flags given to the constructor.
+       *
+       * @param[in] ode_model functor.
+       * @param[in] y0 initial state of the base ode.
+       * @param[in] t0 initial time of the base ode.
+       * @param[in] initial_var flag if sensitivities of initals are needed
+       * @param[in] param_var flag if sensitivities of initals are needed
+       * @param[in] rel_tol Relative tolerance of solver.
+       * @param[in] abs_tol Absolute tolerance of solver.
+       * @param[in] max_num_steps Maximum number of solver steps.
+       * @param[in] solver used solver (0=non-stiff, 1=stiff, 2=stiff
+       * with STALD)
+       */
+      cvodes_integrator(const ode_model<F>& ode_model,
+                        const std::vector<double>& y0,
+                        double t0,
+                        bool initial_var,
+                        bool param_var,
+                        double rel_tol,
+                        double abs_tol,
+                        long int max_num_steps,  // NOLINT(runtime/int)
+                        size_t solver)
+        : ode_model_(ode_model),
+          y0_dbl_(y0),
+          t0_(t0),
+          initial_var_(initial_var),
+          param_var_(param_var),
+          M_(ode_model.size_param()),
+          N_(y0.size()),
+          S_((initial_var ? N_ : 0) + (param_var ? M_ : 0)),
+          size_(N_ * (S_+1)),
+          param_var_ind_(initial_var ? N_ : 0),
+          cvode_mem_(NULL),
+          state_(y0),
+          cvode_state_(N_VMake_Serial(N_,
+                                      &state_[0])),
+          cvode_state_sens_(NULL) {
+        stan::math::check_bounded("cvodes_integrator",
+                                  "solver", solver,
+                                  static_cast<size_t>(0),
+                                  static_cast<size_t>(2));
+
+        // Instantiate CVode memory
+        if (solver == 0) {
+          cvode_mem_ = CVodeCreate(CV_ADAMS, CV_FUNCTIONAL);
```

Comment:
If the object created is expensive to create *and* has reusable
memory, it would make sense to have a singleton implementation.
Otherwise, there's really no point.   

Alternatively, we could enforce good C++ hygiene by using
a pointer-to-implementation (pimpl) behind a resource-allocation-is-init
(RAII) pattern, the same way that std::vector and Eigen::Matrix do.


> On Apr 20, 2016, at 11:17 AM, Daniel Lee <notifications@github.com> wrote:
> 
> In stan/math/rev/mat/functor/cvodes_integrator.hpp:
> 
> > +          S_((initial_var ? N_ : 0) + (param_var ? M_ : 0)),
> > +          size_(N_ * (S_+1)),
> > +          param_var_ind_(initial_var ? N_ : 0),
> > +          cvode_mem_(NULL),
> > +          state_(y0),
> > +          cvode_state_(N_VMake_Serial(N_,
> > +                                      &state_[0])),
> > +          cvode_state_sens_(NULL) {
> > +        stan::math::check_bounded("cvodes_integrator",
> > +                                  "solver", solver,
> > +                                  static_cast<size_t>(0),
> > +                                  static_cast<size_t>(2));
> > +
> > +        // Instantiate CVode memory
> > +        if (solver == 0) {
> > +          cvode_mem_ = CVodeCreate(CV_ADAMS, CV_FUNCTIONAL);
> 
> For a future pull request, but I'm curious; is there a way to avoid calling CVodeCreate and CVodeFree again and again? If we made each solver its own class, could we possibly have some flyweight implementation?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>